Automattic/wp-calypso
# Marketplace: Add Support for Preinstalled Premium Plugins

#### Proposed Changes

* Add support for preinstalled premium plugins (namely: Jetpack Search) to the Marketplace.

This is an alternative, generalized approach to #64828.

The main problem of the previous PR was overthinking the many different possible versions of Jetpack Search. After an internal discussion (pdh6GB-1iz-p2), I've decided to remove the deprecated versions from the equation, consolidating the system into two different items, both requiring a paid upgrade:
- The preinstalled version on Simple sites.
- The standalone version on Jetpack and Atomic.

##### What factors I've removed?

- Classic Search is not considered anymore: users with Classic Search that stumble upon Jetpack Search in the Marketplace will be prompted to upgrade just like anyone else.
- Jetpack and Atomic now work exactly in the same way, both using the standalone version — despite it being technically preinstalled on Atomic.
  - Installing the standalone version on Atomic doesn't create any conflicts (bugs notwithstanding), and at the same time contributes to boosting the active installation count on WPORG, which doesn't consider the WPCOM preinstalled version.

##### Anything else of note?

- Compared to the previous version (and what I said in the internal discussion), now Jetpack and Atomic sites have a 1-click process for "purchase -> install -> activate".
- The pricing support both monthly and yearly variations. 
- This approach is ready for other (internal) preinstalled/premium/standalone features/plugins which might or might not exist in the future.
- They also work with the redesigned plugin layout (see `plugins/plugin-details-layout` feature flag).
- I've intentionally not handled the plugin page's site list, as that will require something more clever.

##### List of changes for Simple sites

| Upgraded? | List | Page |
| - | - | - |
| No | <img width="526" alt="Screenshot 2022-07-01 at 12 49 36" src="https://user-images.githubusercontent.com/2070010/176889058-72d1d0c3-bfec-4c1f-b4b1-2116541fb43c.png"> | <img width="1070" alt="Screenshot 2022-07-01 at 12 47 51" src="https://user-images.githubusercontent.com/2070010/176888800-17c3a01d-8e52-46f5-9657-476519235353.png"> |
| Yes | <img width="525" alt="Screenshot 2022-07-01 at 12 50 00" src="https://user-images.githubusercontent.com/2070010/176889096-ef0ed442-e6cb-4d17-8ec0-e2b12d6f0be7.png"> | <img width="1068" alt="Screenshot 2022-07-01 at 12 47 44" src="https://user-images.githubusercontent.com/2070010/176888817-bac2e482-85d6-4b8d-b3ba-5f242821266b.png"> | 

##### List of changes for Jetpack and Atomic sites

| Installed? | Active? | Upgraded? | List | Page |
| - | - | - | - | - |
| No | No | No | <img width="528" alt="list-not-installed" src="https://user-images.githubusercontent.com/2070010/176901063-e1d15f76-4184-4d3b-8868-f28067ffa51d.png"> | <img width="1066" alt="page-not-installed" src="https://user-images.githubusercontent.com/2070010/176901487-8343276f-6773-4d94-91d1-56f86d69606d.png"> |
| Yes | No | No | <img width="527" alt="list-not-upgraded" src="https://user-images.githubusercontent.com/2070010/176901213-a74f7fa2-8487-42c9-b74a-44631203368d.png"> | <img width="1067" alt="page-not-upgraded" src="https://user-images.githubusercontent.com/2070010/176901590-d7b8050e-ce19-4c15-bb98-989880f66fb7.png"> |
| Yes | Yes | No | <img width="527" alt="list-not-upgraded" src="https://user-images.githubusercontent.com/2070010/176901213-a74f7fa2-8487-42c9-b74a-44631203368d.png"> | <img width="1067" alt="page-not-upgraded" src="https://user-images.githubusercontent.com/2070010/176901590-d7b8050e-ce19-4c15-bb98-989880f66fb7.png"> |
| Yes | No | Yes | <img width="527" alt="list-upgraded-inactive" src="https://user-images.githubusercontent.com/2070010/176901345-7bfa1310-6b85-432f-8f86-6aae491dc5a7.png"> | <img width="1058" alt="page-upgraded-inactive" src="https://user-images.githubusercontent.com/2070010/176901742-4d60f157-62e5-4e6e-9510-24f06b3ba339.png"> |
| Yes | Yes | Yes | <img width="529" alt="list-installed-upgraded" src="https://user-images.githubusercontent.com/2070010/176901371-18c39772-d049-417f-8969-1f47ff32546d.png"> | <img width="1068" alt="page-installed-upgraded" src="https://user-images.githubusercontent.com/2070010/176901416-d488b9a2-86d5-415d-96ff-5752813401a7.png"> |


#### Testing Instructions



* Prepare Simple, Atomic, and Jetpack self-hosted test sites.
* Navigate to the Plugins section, and search for Jetpack Search.
* The plugin page's CTA should be:
  * Simple site without Jetpack Search upgraded:
    * "Jetpack Search is automatically managed for you."
    * [Upgrade Jetpack Search] sending to `/checkout/SITE/jetpack_search_monthly`.
  * Simple site with Jetpack Search upgraded:
    * "Jetpack Search is automatically managed for you."
  * Atomic/Jetpack site without the Jetpack Search plugin installed:
    * [Purchase and activate] sending to `/checkout/SITE/jetpack_search_monthly?redirect_to=/marketplace/jetpack-search/install/SITE`.
  * Atomic/Jetpack with the Jetpack Search plugin installed but not upgraded:
    * [Upgrade Jetpack Search] sending to `/checkout/SITE/jetpack_search_monthly`.
  * Atomic/Jetpack with the Jetpack Search plugin installed and upgraded:
    * Nothing.



Related to #62025
